File: packages/platform/src/ipc/card.ts
/**
 * 卡券管理 IPC 封装 — 卡券 CRUD / 启用状态。
 *
 * 后端：壳层 `commands/card.rs`（InMemoryCardStore + app::card::CardService）。
 *
 * @author agent
 * @created 2026-08-13
 */

import { call } from "./invoke";

/** 卡券来源（与 Rust `CardSource` 对齐）。 */
export type CardSource = "own" | "dock_l1" | "dock_l2";

/** 卡券（与 Rust `Card` 对齐）。 */
export interface Card {
  id: number;
  owner_id: number;
  account_id: string;
  name: string;
  /** text / data / api / image。 */
  card_type: string;
  source: CardSource;
  enabled: boolean;
  text_content: string;
  data_content: string;
  image_url: string;
  image_urls: string;
  api_config: string;
  delay_seconds: number;
  description: string;
}

/** 卡券列表查询入参。 */
export interface CardListRequest {
  owner_id: number;
  page: number;
  page_size: number;
  keyword?: string;
  card_type?: string;
}

/** 查询卡券列表（返回 [列表, 总数]）。 */
export function cardList(request: CardListRequest): Promise<[Card[], number]> {
  return call<[Card[], number]>("card_list", {
    request: {
      owner_id: request.owner_id,
      page: request.page,
      page_size: request.page_size,
      keyword: request.keyword ?? "",
      card_type: request.card_type ?? "",
    },
  });
}

/** 新建卡券。 */
export function cardCreate(ownerId: number, card: Card): Promise<Card> {
  return call<Card>("card_create", { ownerId, card });
}

/** 更新卡券。 */
export function cardUpdate(ownerId: number, card: Card): Promise<void> {
  return call<void>("card_update", { ownerId, card });
}

/** 切换卡券启用状态。 */
export function cardSetEnabled(ownerId: number, cardId: number, enabled: boolean): Promise<void> {
  return call<void>("card_set_enabled", {
    request: { owner_id: ownerId, card_id: cardId, enabled },
  });
}

/** 删除卡券。 */
export function cardDelete(ownerId: number, cardId: number): Promise<void> {
  return call<void>("card_delete", { request: { owner_id: ownerId, card_id: cardId } });
}

File: business/src/delivery/execution/card.rs
//! 卡券模型与来源优先级选择。
//!
//! 对齐 Python 版 `_auto_delivery` 的卡券匹配逻辑：
//! 按来源优先级（own 自有 → dock_l1 一级对接 → dock_l2 二级对接）分组，
//! 取第一个"恰好 1 张卡"的分组；同组多张或全组为空则无匹配。

use serde::{Deserialize, Serialize};

/// 卡券来源。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CardSource {
    /// 自有卡券。
    Own,
    /// 一级对接。
    DockL1,
    /// 二级对接。
    DockL2,
}

impl CardSource {
    pub fn as_str(&self) -> &'static str {
        match self {
            CardSource::Own => "own",
            CardSource::DockL1 => "dock_l1",
            CardSource::DockL2 => "dock_l2",
        }
    }

    #[allow(clippy::should_implement_trait)]
    pub fn from_str(value: &str) -> Self {
        match value {
            "dock_l1" => CardSource::DockL1,
            "dock_l2" => CardSource::DockL2,
            _ => CardSource::Own,
        }
    }
}

/// 来源优先级（own 最高）。
const SOURCE_PRIORITY: [CardSource; 3] = [CardSource::Own, CardSource::DockL1, CardSource::DockL2];

/// 卡券（业务层从存储加载）。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Card {
    #[serde(default)]
    pub id: i64,
    /// 所属用户（管理视角）。
    #[serde(default)]
    pub owner_id: i64,
    /// 所属账号（管理视角）。
    #[serde(default)]
    pub account_id: String,
    pub name: String,
    /// text / data / api / image。
    pub card_type: String,
    #[serde(default = "default_source")]
    pub source: CardSource,
    /// 是否启用（自动发货可选）。
    #[serde(default = "default_true")]
    pub enabled: bool,
    /// 固定文字内容（text 类型）。
    #[serde(default)]
    pub text_content: String,
    /// 批量数据内容（data 类型，多行一条）。
    #[serde(default)]
    pub data_content: String,
    /// 图片 URL（image 类型）。
    #[serde(default)]
    pub image_url: String,
    /// 多图片 URL 列表（JSON 数组字符串）。
    #[serde(default)]
    pub image_urls: String,
    /// API 拉取配置（JSON 字符串）。
    #[serde(default)]
    pub api_config: String,
    /// 发货延时（秒）。
    #[serde(default)]
    pub delay_seconds: u32,
    /// 备注（追加到内容后）。
    #[serde(default)]
    pub description: String,
}

fn default_source() -> CardSource {
    CardSource::Own
}

fn default_true() -> bool {
    true
}

impl Card {
    /// 是否为对接卡券（card_only 模式下跳过）。
    pub fn is_dock(&self) -> bool {
        matches!(self.source, CardSource::DockL1 | CardSource::DockL2)
    }
}

/// 卡券选择器 — 来源优先级唯一匹配。
pub struct CardSelector;

impl CardSelector {
    /// 按来源优先级选取唯一卡券。
    ///
    /// 返回 `(卡券, 来源)`；无唯一匹配返回 `None`。
    pub fn select(cards: &[Card]) -> Option<&Card> {
        SOURCE_PRIORITY
            .iter()
            .filter_map(|source| {
                let group: Vec<&Card> =
                    cards.iter().filter(|card| card.source == *source).collect();
                match group.len() {
                    1 => Some(group[0]),
                    _ => None, // 0 张或 >1 张均跳过该来源
                }
            })
            .next()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn card(source: CardSource) -> Card {
        Card {
            id: 1,
            owner_id: 0,
            account_id: String::new(),
            name: "卡".to_string(),
            card_type: "text".to_string(),
            source,
            enabled: true,
            text_content: "内容".to_string(),
            data_content: String::new(),
            image_url: String::new(),
            image_urls: String::new(),
            api_config: String::new(),
            delay_seconds: 0,
            description: String::new(),
        }
    }

    #[test]
    fn selects_own_first() {
        let cards = vec![card(CardSource::DockL1), card(CardSource::Own)];
        assert_eq!(
            CardSelector::select(&cards).map(|c| c.source),
            Some(CardSource::Own)
        );
    }

    #[test]
    fn skips_source_with_multiple_cards() {
        // own 有 2 张 → 跳过，落到唯一 1 张的 dock_l1。
        let cards = vec![
            card(CardSource::Own),
            card(CardSource::Own),
            card(CardSource::DockL1),
        ];
        assert_eq!(
            CardSelector::select(&cards).map(|c| c.source),
            Some(CardSource::DockL1)
        );
    }

    #[test]
    fn none_when_no_unique_group() {
        let cards = vec![card(CardSource::Own), card(CardSource::Own)];
        assert!(CardSelector::select(&cards).is_none());
        assert!(CardSelector::select(&[]).is_none());
    }

    #[test]
    fn dock_detection() {
        assert!(card(CardSource::DockL1).is_dock());
        assert!(card(CardSource::DockL2).is_dock());
        assert!(!card(CardSource::Own).is_dock());
    }

    #[test]
    fn source_str_roundtrip() {
        assert_eq!(CardSource::from_str("dock_l1"), CardSource::DockL1);
        assert_eq!(CardSource::from_str("unknown"), CardSource::Own);
        assert_eq!(CardSource::Own.as_str(), "own");
    }
}
